ext/sqlite3/sqlite3.c

```diff
@@ -1395,6 +1395,127 @@ PHP_METHOD(sqlite3stmt, readOnly)
 }
 /* }}} */
 
+static int php_sqlite3_bind_params(php_sqlite3_stmt *stmt_obj) /* {{{ */
+{
+	struct php_sqlite3_bound_param *param;
+
+	if (stmt_obj->bound_params) {
+		ZEND_HASH_FOREACH_PTR(stmt_obj->bound_params, param) {
+			zval *parameter;
+			/* parameter must be a reference? */
+			if (Z_ISREF(param->parameter)) {
+				parameter = Z_REFVAL(param->parameter);
+			} else {
+				parameter = &param->parameter;
+			}
+
+			/* If the ZVAL is null then it should be bound as that */
+			if (Z_TYPE_P(parameter) == IS_NULL) {
+				sqlite3_bind_null(stmt_obj->stmt, param->param_number);
+				continue;
+			}
+
+			switch (param->type) {
+				case SQLITE_INTEGER:
+					convert_to_long(parameter);
+#if ZEND_LONG_MAX > 2147483647
+					sqlite3_bind_int64(stmt_obj->stmt, param->param_number, Z_LVAL_P(parameter));
+#else
+					sqlite3_bind_int(stmt_obj->stmt, param->param_number, Z_LVAL_P(parameter));
+#endif
+					break;
+
+				case SQLITE_FLOAT:
+					convert_to_double(parameter);
+					sqlite3_bind_double(stmt_obj->stmt, param->param_number, Z_DVAL_P(parameter));
+					break;
+
+				case SQLITE_BLOB:
+				{
+					php_stream *stream = NULL;
+					zend_string *buffer = NULL;
+					if (Z_TYPE_P(parameter) == IS_RESOURCE) {
+						php_stream_from_zval_no_verify(stream, parameter);
+						if (stream == NULL) {
+							php_sqlite3_error(stmt_obj->db_obj, "Unable to read stream for parameter %ld", param->param_number);
+							return 0;
+						}
+						buffer = php_stream_copy_to_mem(stream, PHP_STREAM_COPY_ALL, 0);
+					} else {
+						buffer = zval_get_string(parameter);
+					}
+
+					if (buffer) {
+						sqlite3_bind_blob(stmt_obj->stmt, param->param_number, ZSTR_VAL(buffer), ZSTR_LEN(buffer), SQLITE_TRANSIENT);
+						zend_string_release_ex(buffer, 0);
+					} else {
+						sqlite3_bind_null(stmt_obj->stmt, param->param_number);
+					}
+					break;
+				}
+
+				case SQLITE3_TEXT:
+					convert_to_string(parameter);
+					sqlite3_bind_text(stmt_obj->stmt, param->param_number, Z_STRVAL_P(parameter), Z_STRLEN_P(parameter), SQLITE_STATIC);
+					break;
+
+				case SQLITE_NULL:
+					sqlite3_bind_null(stmt_obj->stmt, param->param_number);
+					break;
+
+				default:
+					php_sqlite3_error(stmt_obj->db_obj, "Unknown parameter type: %pd for parameter %pd", param->type, param->param_number);
+					return 0;
+			}
+		} ZEND_HASH_FOREACH_END();
+	}
+	return 1;
+}
+/* }}} */
+
+
+/* {{{ proto string SQLite3Stmt::getSQL([expanded = false])
+   Returns the SQL statement used to prepare the query. If expanded is true, binded parameters and values will be expanded. */
+PHP_METHOD(sqlite3stmt, getSQL)
+{
+	php_sqlite3_stmt *stmt_obj;
+	zend_bool expanded = 0;
+	zval *object = getThis();
+	stmt_obj = Z_SQLITE3_STMT_P(object);
+	int rc;
+
+	if (zend_parse_parameters(ZEND_NUM_ARGS(), "|b", &expanded) == FAILURE) {
+		return;
+	}
+
+	SQLITE3_CHECK_INITIALIZED(stmt_obj->db_obj, stmt_obj->initialised, SQLite3);
```

Comment:
This looks wrong (as in several other places of the existing code as well). We' re mixing the initialization check of `SQLite3` and `SQLite3Stmt` here.

Reply:
I just copy-pasted existing code here yeah, what would you suggest instead?

Reply:
Not sure (yet) what's the proper way to do it. Frankly, I'd rather wish that we didn't have to do these initialization checks at all.

Reply:
Well, I think it's fine to leave this for now, and to review all those checks later. :)